composer/datasets/streaming/download.py

```diff
@@ -52,100 +33,68 @@ def download_from_s3(remote: str, local: str, timeout: float) -> None:
+        max_retries (int, default 2): Number of download re-attempts before giving up.
+        timeout (float, default 60): How long to wait for file to download before raising an exception.
     """
-    # If we already have the file cached locally, we are done.
-    if os.path.exists(local):
-        return
-
-    # Check if there is a tmp file.
-    local_tmp = local + '.tmp'
-    if os.path.exists(local_tmp):
-        # Get tmp file created time
-        local_tmp_create_time = os.path.getctime(local_tmp)
-
-        # Get current disk time, more consistent than system time
-        with tempfile.NamedTemporaryFile() as f:
-            current_disk_time = os.path.getctime(f.name)
-
-        if current_disk_time - local_tmp_create_time < timeout + 1:  # 1s buffer to avoid race condition
-            # If the tmp file is recent, it is either (1) from a very recent crashed run, or (2) another thread is actively downloading it.
-            # So we wait but don't error out, in case we are in situation (1)
-            try:
-                wait_for_download(local, timeout)
-                return
-            except TimeoutError:
-                pass
-
-        # The tmp file is old, it is either (1) from a crashed run or (2) another thread is downloading it but is taking too long, and will timeout.
-        # Let's delete the tmp file. If situation (1), this is safe. If situation (2), the other thread is expected to crash with a TimeoutError anyways, so this is fine.
+    ok = False
+    error_msgs = []
+    for _ in range(1 + max_retries):
         try:
-            os.remove(local_tmp)
-        except OSError:
-            # This occurs if another download thread got to the delete first.
-            pass
-
-    # There is no tmp file, so attempt to make it.
-    # If this fails, another download thread beat us to it, so wait.
-    local_dir = os.path.dirname(local)
-    os.makedirs(local_dir, exist_ok=True)
-    try:
-        with open(local_tmp, 'xb') as out:
-            out.write(b'')
-    except FileExistsError:
-        # If we run out of time here, we know a download thread was active and exceeded timeout, so we should error out.
-        wait_for_download(local, timeout)
-        return
-
-    # We succesfully created the tmp file. Perform the download and rename.
-    download(remote, local_tmp, timeout)
-    os.rename(local_tmp, local)
+            if wait:
+                start = time.time()
+                while not os.path.exists(local):
+                    if time.time() - start > timeout:
+                        raise TimeoutError(f"Waited longer than {timeout}s for other worker to download {local}.")
+                    time.sleep(0.25)
+            else:
+                dispatch_download(remote, local, timeout=timeout)
+            ok = True
+            break
+        except Exception as e:  # Retry for all causes of failure.
```

Comment:
It seems like this catch is overly broad because we don't know what errors the underlying download handler will be throwing. Like it may be more than just `TimeoutErrors`.

We should have a refactor coming in soon so that `dispatch_download` gets refactored to `composer.utils.file_helpers`. At that time I think we will have a bit more control over the errors and can make this exception catching stricter. Let me know if that's OK.

Reply:
OK, fair point -- especially if we switch to libcloud, then the error handling might be different. But it does appear boto3 has some nice handling: https://boto3.amazonaws.com/v1/documentation/api/latest/guide/error-handling.html

Reply:
Punting this as `libcloud` refactor should be imminent.